subsys/settings/Kconfig

```diff
@@ -6,88 +6,137 @@
 	  It supports several back-ends to store and load serialized data from
 	  and it can do so atomically for all involved modules.
 
-# Hidden option to enable encoding length into settings entry
-config SETTINGS_ENCODE_LEN
+config SETTINGS_RUNTIME
+	bool "Runtime support"
+	default y
 	depends on SETTINGS
-	bool
+	help
+	  Enables runtime support.
+
+config SETTINGS_FCB
+	bool "FCB non-volatile storage support"
+	default n
+	depends on SETTINGS
+	help
+	  Enables FCB storage support
 
-config SETTINGS_USE_BASE64
-	bool "encoding value using base64"
+config SETTINGS_FS
+	bool "FILE non-volatile storage support"
+	default n
 	depends on SETTINGS
-	select BASE64
 	help
-	  Enables values encoding using Base64.
+	  Enables file storage support
+
+config SETTINGS_NVS
+	bool "NVS non-volatile storage support"
+	default n
+	depends on SETTINGS
+	help
+	  Enables NVS storage support
 
 choice
-	prompt "Storage back-end"
-	default SETTINGS_FCB if FCB
+	prompt "Default backend type"
+	default SETTINGS_DEFAULT_FCB
 	depends on SETTINGS
 	help
-	  Storage back-end to be used by the settings subsystem.
+	  Default storage back-end to be used by the settings subsystem.
 
-config SETTINGS_FCB
+config SETTINGS_DEFAULT_FCB
 	bool "FCB"
 	depends on FCB
+	select SETTINGS_FCB
 	help
-	  Use FCB as a settings storage back-end.
+	  Use FCB as the default storage back-end.
 
-config SETTINGS_FS
+config SETTINGS_DEFAULT_FS
 	bool "File System"
 	depends on FILE_SYSTEM
-	select SETTINGS_ENCODE_LEN
+	select SETTINGS_FS
 	help
-	  Use a file system as a settings storage back-end.
+	  Use a file system as the default storage back-end.
+
+config SETTINGS_DEFAULT_NVS
+	bool "NVS"
+	depends on NVS
+	select SETTINGS_NVS
+	help
+	  Use NVS as the default storage back-end.
+
+config SETTINGS_DEFAULT_NONE
+	bool "None"
+	help
+	  Don't use a default storage back-end.
 endchoice
 
-config SETTINGS_FCB_NUM_AREAS
+config SETTINGS_DEFAULT_FCB_NUM_AREAS
 	int "Number of flash areas used by the settings subsystem"
 	default 8
-	depends on SETTINGS && SETTINGS_FCB
+	depends on SETTINGS && SETTINGS_DEFAULT_FCB
 	help
 	  Number of areas to allocate in the settings FCB. A smaller number is
 	  used if the flash hardware cannot support this value.
 
-config SETTINGS_FCB_MAGIC
+config SETTINGS_DEFAULT_FCB_MAGIC
 	hex "FCB magic for the settings subsystem"
 	default 0xc0ffeeee
-	depends on SETTINGS && SETTINGS_FCB
+	depends on SETTINGS && SETTINGS_DEFAULT_FCB
 	help
 	  Magic 32-bit word for to identify valid settings area
 
-config SETTINGS_FCB_FLASH_AREA
+config SETTINGS_DEFAULT_FCB_FLASH_AREA
 	int "Flash area id used for settings"
 	default 4
-	depends on SETTINGS && SETTINGS_FCB
+	depends on SETTINGS && SETTINGS_DEFAULT_FCB
 	help
 	  Id of the Flash area where FCB instance used for settings is
 	  expected to operate.
 
-config SETTINGS_FS_DIR
-	string "Serialization directory"
-	default "/settings"
-	depends on SETTINGS && SETTINGS_FS
+config SETTINGS_DEFAULT_FS_MNT
```

Comment:
No, this should be the mount point, when trying to figure out the old settings with nffs I had a lot of difficulty finding out where the mount point should be. By defining it like this it is possible to define nffs with the mount point and then later use this mount point for your file.

Reply:
I typical situation settings will coexisting with other modules in the FS. Otherwise doesn't make sense to us FS backend at all (because NVS and FCB is much smaller and faster tha NFFS) . FS should be mounted before settings initialization. So this should be directory.

Reply:
Please have a good look at this, I don't think the current definitions make any sense. The problem is that the filename and filedir should contain the mount point. I have no idea how to solve this, except by specifying the mount point in the config file. Maybe you have a better idea, but according to me the current kconfig does not work.

Reply:
> The problem is that the filename and filedir should contain the mount point.

Yes - and it should be like that. Settings need absolute path to the file. Directory is place when we can create temporary file while compressing the settings file. I just check that old file name and directory name didn't contain mount points - it was jut because multi-fs was introduced after FS back-end. (i think this might have mislead you)

Reply:
OK, the reworked settings does not need a directory name, so there is no need to define it.
Regarding the mount point: when the default backend is selected to be of type file everything should work regardless of how someone wishes to use it. When the correct mount point for the app is selected the settings ```CONFIG_SETTINGS_DEFAULT_FS_MNT``` and the ```CONFIG_SETTINGS_DEFAULT_FILENAME``` can be changed to reflect the chosen mount point.
I am open to a proposal that changes this, but it should:
a. Work out of the box for a sample,
b. Be easily adapted to user requirements.

Reply:
I didn't catch that this is for sample only, and FS backend needs only SETTINGS_DEFAULT_FS_FILE.
Pleas remove this kconfig form here - as this is not for the subsys but for samles.

Reply:
This is not really for sample only, maybe I should redefine the SETTINGS_DEFAULT_FS_FILE as the concat of SETTINGS_DEFAULT_FS_MNT and SETTINGS_DEFAULT_FS_FILE (without the mount point). Kconfig can then be used to set SETTINGS_DEFAULT_FS_MNT according to the user needs.

Reply:
Or keep only SETTINGS_DEFAULT_FS_FILE with absolute path. In sames extract the mount point from it. Advantage is less properties to see for the end-user. I thin it is also intuitive.

Reply:
Ok - lets do this as you proposed (SETTINGS_DEFAULT_FS_MNT and SETTINGS_DEFAULT_FS_FILE )